numalogic/registry/redis_registry.py

```diff
@@ -0,0 +1,247 @@
+                raise ValueError("Either One of 'latest' or 'version' needed in load method call")
+            model_key = self.__get_model_key(production, version, skeys, dkeys)
+            serialized_model, model_version, model_timestamp = self.client.hmget(
+                name=model_key, keys=["model", "version", "timestamp"]
+            )
+            deserialized_model = loads(serialized_model)
+            metadata = self.__load_metadata(model_key)
+            return ArtifactData(
+                artifact=deserialized_model,
+                metadata=metadata,
+                extras={
+                    "model_timestamp": int(model_timestamp.decode()),
+                    "model_version": model_version.decode(),
+                },
+            )
+        except ModelKeyNotFound as model_key_error:
+            _LOGGER.exception("Missing Key: %s", model_key_error)
+            return None
+        except RedisError as ex:
+            _LOGGER.exception("Unexpected error: %s", ex)
+            return None
+
+    def save(
+        self,
+        skeys: Sequence[str],
+        dkeys: Sequence[str],
+        artifact: Artifact,
+        **metadata: str,
+    ) -> Optional[str]:
+        """
+        Saves the artifact into redis registry and updates version.
+        Args:
+            skeys: static key fields as list/tuple of strings
+            dkeys: dynamic key fields as list/tuple of strings
+            artifact: primary artifact to be saved
+            metadata: additional metadata surrounding the artifact that needs to be saved
+
+        Returns:
+            model version
+        """
+        try:
+            production_key = self.construct_key(skeys, dkeys, ["PRODUCTION"])
+            version = 0
+            if self.client.exists(production_key):
+                _LOGGER.info("Production key exists for the model")
+                version_key = self.client.get(name=production_key)
+                version = int(self.get_version(version_key.decode())) + 1
+            new_version_key = self.__save_artifact(artifact, skeys, dkeys, str(version))
+            if metadata:
+                self.__save_metadata(metadata, new_version_key)
+            self.client.expire(name=new_version_key, time=self.ttl)
+            _LOGGER.info("Model is successfully with the key = %s", new_version_key)
+            return str(version)
+        except RedisError as ex:
+            _LOGGER.exception("Unexpected error: %s", ex)
+            return None
+
+    def delete(self, skeys: Sequence[str], dkeys: Sequence[str], version: str) -> None:
+        """
+        Saves the artifact into mlflow registry and updates version.
+        Args:
+            skeys: static key fields as list/tuple of strings
+            dkeys: dynamic key fields as list/tuple of strings
+            version: model version to delete
+        """
+        try:
+            del_key = self.construct_key(skeys, dkeys, [version])
```

Comment:
should we check if the key being deleted is referenced by "production"?

Reply:
Interesting thought but then this is gonna make this api call very expensive. we will have to fetch all the production keys. Then all these production keys will have to be checked for the key that we are trying to delete.